doc/api/async_hooks.md

```diff
@@ -579,6 +579,121 @@ const server = net.createServer((conn) => {
 Promise contexts may not get valid `triggerAsyncId`s by default. See
 the section on [promise execution tracking][].
 
+### Class: `AsyncLocal`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+This class can be used to store a value which follows asynchronous execution
+flow. Any value set on an `AsyncLocal` instance is propagated to any callback
+or promise executed within the flow. Because of that, a continuation local
+storage can be build with an `AsyncLocal` instance. This API is similar to
+thread local storage in other runtimes and languages.
+
+The implementation relies on async hooks to follow the execution flow.
+So, if an application or a library does not play nicely with async hooks,
+the same problems will be seen with the `AsyncLocal` API. In order to fix
+such issues the `AsyncResource` API should be used.
+
+The following example shows how to use `AsyncLocal` to build a simple logger
+that assignes ids to HTTP requests and includes them into messages logged
+within each request.
+
+```js
+const http = require('http');
+const { AsyncLocal } = require('async_hooks');
+
+const asyncLocal = new AsyncLocal();
+
+function print(msg) {
+  const id = asyncLocal.get();
+  console.log(`${id !== undefined ? id : '-'}:`, msg);
+}
+
+let idSeq = 0;
+http.createServer((req, res) => {
+  asyncLocal.set(idSeq++);
+  print('start');
+  setImmediate(() => {
+    print('finish');
+    res.end();
+  });
+}).listen(8080);
+
+http.get('http://localhost:8080');
+http.get('http://localhost:8080');
+// Prints:
+//   0: start
+//   1: start
+//   0: finish
+//   1: finish
+```
+
+#### `new AsyncLocal()`
+
+Creates a new instance of `AsyncLocal`.
+
+### `asyncLocal.get()`
+
+* Returns: {any}
+
+Returns the value of the `AsyncLocal` in current execution context,
+or `undefined` if the value is not set or the `AsyncLocal` was removed.
+
+### `asyncLocal.set(value)`
+
+* `value` {any}
+
+Sets the value for the `AsyncLocal` within current execution context.
+
+Once set, the value will be kept through the subsequent asynchronous calls,
+unless overridden by calling `asyncLocal.set(value)`:
+
+```js
+const asyncLocal = new AsyncLocal();
+
+setImmediate(() => {
+  asyncLocal.set('A');
+
+  setImmediate(() => {
+    console.log(asyncLocal.get());
+    // Prints: A
+
+    asyncLocal.set('B');
+    console.log(asyncLocal.get());
+    // Prints: B
+  });
+
+  console.log(asyncLocal.get());
+  // Prints: A
+
+  // Stop further value propagation
+  asyncLocal.set(undefined);
+
+  console.log(asyncLocal.get());
+  // Prints: undefined
+
+  setImmediate(() => {
+    console.log(asyncLocal.get());
+    // Prints: undefined
+  });
+});
+```
+
+If the `AsyncLocal` was removed before this call is made,
+[`ERR_ASYNC_LOCAL_CANNOT_SET_VALUE`][] is thrown.
```

Comment:
Not sure if it is that nice for users to get an exception in this case. What about silently ignoring calls to `set` and add `asyncLocal.enabled` (or similar)?

Reply:
My idea was to have `.remove()` method as a terminal operation in `AsyncLocal`'s life cycle. That's why I implemented fail fast behavior in `.set()` method.

A pair of `.enable()`/`.disable()` methods, on the other hand would allow to reuse the `AsyncLocal`. But current implementation removes all underlying values. And I'm afraid that some users may expect to see the same value under `AsyncLocal` when they disable and re-enable it.

Do you think it's fine to keep current `.remove()` method's behavior and only to rename it to `.disable()`? Or it's better to get rid of underlying values removal behavior?

Reply:
Additional context. I have implemented `.remove()` method as a terminal operation to force users to create a new `AsyncLocal` instance instead of enabling an existing one (in the `.enable()`/`.disable()` methods case). I believe that such approach will help to avoid various race issues when dealing with the API.

Reply:
Are you expecting users to call `.remove()` in typical scenarios?
I miss the usecase (except dealing with problems like leaks to avoid further breakdowns).
I expect that calls to `.set()` are spread in the whole application therefore you either have to check everywhere if the async local is still alive or use a try/catch.

Reply:
I expect this method to be called in niche scenarios, say, in tests or any situations when `AsyncLocal` needs to temporarily disabled or permanently disposed.

I get your point about the error and agree that it's better to silently ignore `.set()` calls on a removed `AsyncLocal`. Going to do this change.

Considering our discussion, do you still think that switching from `.remove()` to `.enable()`/`.disable()` will improve the API?

Reply:
Shipped no-error `.set()` in https://github.com/nodejs/node/pull/31016/commits/4efc80ff41151b845f78445c535c975600746e63

Reply:
I don't think we should offer `.enable()`. I would vote for `.disable()` or `.destroy()` but others are also ok.

Reply:
Shipped `.remove()` -> `.destroy()` rename in https://github.com/nodejs/node/pull/31016/commits/b65d09a4d67ef063de4b97013ef0da2c89793ab2